packages/datatrak-web/src/api/stream.ts

```diff
@@ -0,0 +1,197 @@
+import { SYNC_STREAM_MESSAGE_KIND } from '@tupaia/constants';
+import { API_URL } from './api';
+
+/** Connect to a streaming endpoint and async yield messages.
+ *
+ * ```js
+ * for await (const { kind, message } of centralServer.stream(() => ({
+ *   endpoint: `some/kind/of/stream`,
+ * }))) {
+ *   switch (kind) {
+ *     case SYNC_STREAM_MESSAGE_KIND.SOMETHING:
+ *       // do something
+ *       break;
+ *     case SYNC_STREAM_MESSAGE_KIND.END:
+ *       // finalise
+ *       break;
+ *     default:
+ *       console.warn(`Unknown message kind: ${kind}`);
+ *   }
+ * }
+ * ```
+ *
+ * The streaming endpoint needs to talk the Tamanu Streaming Protocol: a lightweight framing
+ * protocol which includes a 2-byte Message Kind (unsigned int) and an optional JSON payload.
+ * The stream MUST end with an `END` Message Kind (which may have a payload): if the stream
+ * does not receive an `END` message, it is assumed to be incomplete and is automatically
+ * restarted; this protects against unexpected stream disconnections.
+ *
+ * There are two possible layers of retry logic: on connection, using the endpointFn `options`
+ * map, you can set `backoff` to retry the fetch on initial failure. This applies on top of the
+ * stream retries, controlled by `streamRetryAttempts` (default 10) and `streamRetryInterval`
+ * (milliseconds, default 10 seconds), which will restart the entire stream if it fails early.
+ * Set `streamRetryAttempts` to 1 to disable the retries.
+ *
+ * Because the entire stream is restarted during a stream retry, the endpoint is not a fixed URL
+ * but instead a function which is expected to return an object: `{ endpoint, query, options }`.
+ * The `endpoint` key is required, the others default to `{}` if not present. These are passed to
+ * and interpreted the same as for `.fetch()` above.
+ *
+ * For example, you can track some progress information from the messages you receive, and then
+ * provide a "start from this point" query parameter to the next retry call. This avoids either
+ * receiving the full stream contents again or keeping track of stream session state server-side.
+ *
+ * Message payloads are expected to be JSON, and by default are parsed directly within this
+ * function. If you expect non-JSON payloads, or if you want to obtain the raw payload for some
+ * other reason, pass `decodeMessage: false`. This will be slightly faster as the framing allows
+ * us to seek forward through the received data rather than read every byte.
+ *
+ * @param {() => ({ endpoint: string, method?: 'GET' | 'POST' | 'PUT' | 'DELETE', options?: Record<string, any> })} endpointFn
+ * @param {{ decodeMessage?: boolean, streamRetryAttempts?: number, streamRetryInterval?: number }} streamOptions
+ * @returns
+ */
+export async function* stream(
+  endpointFn,
+  { decodeMessage = true, streamRetryAttempts = 10, streamRetryInterval = 10000 } = {},
+) {
```

Comment:
you can pull the @param typings from the doc comment into actual typescript types